mbf_abstract_nav/include/mbf_abstract_nav/abstract_action.h

```diff
@@ -51,106 +50,109 @@ template <typename Action, typename Execution>
 class AbstractAction
 {
  public:
-  typedef boost::bimaps::bimap<uint8_t, std::string> SlotGoalIdMap;
-
   typedef boost::shared_ptr<AbstractAction> Ptr;
 
   typedef typename actionlib::ActionServer<Action>::GoalHandle GoalHandle;
   typedef boost::function<void (GoalHandle &goal_handle, Execution &execution)> RunMethod;
 
+  typedef struct
+  {
+    typename Execution::Ptr execution;
+    boost::thread* thread_ptr;
+    GoalHandle goal_handle;
+  } ConcurrencySlot;
+
   AbstractAction(
       const std::string& name,
       const RobotInformation &robot_info,
       const RunMethod run_method
   ) : name_(name), robot_info_(robot_info), run_(run_method){}
 
-  void start(
-      GoalHandle goal_handle,
+  virtual void start(
+      GoalHandle &goal_handle,
       typename Execution::Ptr execution_ptr
   )
   {
-    boost::lock_guard<boost::mutex> guard(map_mtx_);
-    typename SlotGoalIdMap::left_const_iterator slot
-        = concurrency_slots_.left.find(goal_handle.getGoal()->concurrency_slot);
-    if(slot != concurrency_slots_.left.end()) // if there is a plugin running on the same slot, cancel it // TODO make thread safe
+    uint8_t slot = goal_handle.getGoal()->concurrency_slot;
+
+    boost::lock_guard<boost::mutex> guard(slots_mtx_);
+    typename std::map<uint8_t, ConcurrencySlot>::iterator slot_it = concurrency_slots_.find(slot);
+    if(slot_it != concurrency_slots_.end())
     {
-      typename std::map<const std::string, const typename Execution::Ptr>::const_iterator elem
-          = executions_.find(slot->second);
-      if(elem != executions_.end())
-      {
-        elem->second->cancel();
-      }
-      concurrency_slots_.left.erase(slot->first);
+      // if there is a plugin running on the same slot, cancel it
+      slot_it->second.execution->cancel(); // TODO make thread safe
```

Comment:
Still not thread safe, should we do that in another PR or commit? `cancel` could reach the `CleanUp` part of `runAndCleanUp` after assigning the new thread pointer in line 88. This will be the case, since the mutex blocks until we finished assigning. Then it deletes the new thread instead of the old one, or am I wrong?

Reply:
Probably yes,,,  not thread safe. But I would rather fix in a separated PR, as this is already quite complex